src/sender.ts

```diff
@@ -0,0 +1,1105 @@
+ * Detailed description of the Sender's configuration options can be found in
+ * the <a href="SenderOptions.html">SenderOptions</a> documentation.
+ * </p>
+ * <p>
+ * Extra options can be provided to the Sender in the <i>extraOptions</i> configuration object. <br>
+ * A custom logging function and a custom HTTP(S) agent can be passed to the Sender in this object. <br>
+ * The logger implementation provides the option to direct log messages to the same place where the host application's
+ * log is saved. The default logger writes to the console. <br>
+ * The custom HTTP(S) agent option becomes handy if there is a need to modify the default options set for the
+ * HTTP(S) connections. A popular setting would be disabling persistent connections, in this case an agent can be
+ * passed to the Sender with <i>keepAlive</i> set to <i>false</i>. <br>
+ * For example: <i>Sender.fromConfig(`http::addr=host:port`, { agent: new undici.Agent({ connect: { keepAlive: false } })})</i> <br>
+ * If no custom agent is configured, the Sender will use its own agent which overrides some default values
+ * of <i>undici.Agent</i>. The Sender's own agent uses persistent connections with 1 minute idle timeout, pipelines requests default to 1.
+ * </p>
+ */
+class Sender {
+  /** @private */ static DEFAULT_HTTP_AGENT;
+  /** @private */ static DEFAULT_HTTPS_AGENT;
+
+  /** @private */ http; // true if the protocol is HTTP/HTTPS, false if it is TCP/TCPS
+  /** @private */ secure; // true if the protocol is HTTPS or TCPS, false otherwise
+  /** @private */ host;
+  /** @private */ port;
+
+  /** @private */ socket;
+
+  /** @private */ username;
+  /** @private */ password;
+  /** @private */ token;
+
+  /** @private */ tlsVerify;
+  /** @private */ tlsCA;
+
+  /** @private */ bufferSize;
+  /** @private */ maxBufferSize;
+  /** @private */ buffer;
+  /** @private */ toBuffer;
+  /** @private */ doResolve;
+  /** @private */ position;
+  /** @private */ endOfLastRow;
+
+  /** @private */ autoFlush;
+  /** @private */ autoFlushRows;
+  /** @private */ autoFlushInterval;
+  /** @private */ lastFlushTime;
+  /** @private */ pendingRowCount;
+
+  /** @private */ requestMinThroughput;
+  /** @private */ requestTimeout;
+  /** @private */ retryTimeout;
+
+  /** @private */ hasTable;
+  /** @private */ hasSymbols;
+  /** @private */ hasColumns;
+
+  /** @private */ maxNameLength;
+
+  /** @private */ log;
+  /** @private */ agent;
+  /** @private */ jwk;
+
+  /**
+   * Creates an instance of Sender.
+   *
+   * @param {SenderOptions} options - Sender configuration object. <br>
+   * See SenderOptions documentation for detailed description of configuration options. <br>
+   */
+  constructor(options: SenderOptions) {
+    if (!options || !options.protocol) {
+      throw new Error("The 'protocol' option is mandatory");
+    }
+    replaceDeprecatedOptions(options);
+
+    this.log = typeof options.log === "function" ? options.log : log;
```

Comment:
instead of allowing people to use the custom logger can we have the ability to set the verbosity of it? i guess that is a lot better approach since most of us just want to see errors

Reply:
In my case i find this interesting, for example we use Pino instead of console.log, so it's great to allow overriding the default logger

Reply:
indeed, but if you do not want to use a custom logger you are stuck with the info level of the default logger which is quite spammy on console output. The way how I see it, is why have the levels of the default logger if one cannot change them